File: src/main/java/conservative/DenominationSupplier.java
package conservative;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class DenominationSupplier {

    enum SortOrder {
        ASC, DESC
    }
    public static List<Models.Denomination> forCurrency(String currency, SortOrder sortOrder) {
        List<Models.Denomination> denominations = switch (currency) {
            case "USD" -> Arrays.stream(Models.Dollar.values()).collect(Collectors.toList());
            case "EUR" -> Arrays.stream(Models.Euro.values()).collect(Collectors.toList());
            default -> throw new UnsupportedOperationException("Uh Oh!");
        };

        Comparator<Models.Denomination> ascByValue = Comparator.comparingInt(Models.Denomination::getValue);
        Comparator<Models.Denomination> descByValue = ascByValue.reversed();

        if (sortOrder == SortOrder.ASC) {
            return denominations.stream().sorted(ascByValue).collect(Collectors.toList());
        } else {
            return denominations.stream().sorted(descByValue).collect(Collectors.toList());
        }
    }
}

File: src/main/kotlin/progressive/DenominationSupplier.kt
package progressive

class DenominationSupplier {

    enum class SortOrder {
        ASC, DESC
    }
    companion object {
        fun forCurrency(currency: String, sortOrder: SortOrder? = SortOrder.DESC): List<Denomination> {
            return when (currency) {
                "USD" -> Dollar.entries
                "EUR" -> Euro.entries
                else -> throw UnsupportedOperationException("Unsupported currency: $currency")
            }.let { denominations ->
                when (sortOrder) {
                    SortOrder.DESC -> denominations.sortedByDescending { it.value }
                    SortOrder.ASC -> denominations.sortedBy { it.value }
                    else -> throw UnsupportedOperationException("Unsupported sort order: $sortOrder")
                }
            }
        }
    }
}
